nova/src/ccs/mod.rs

```diff
@@ -0,0 +1,532 @@
+pub struct CCSShape<G: CurveGroup> {
+    /// `m` in the CCS/HyperNova papers.
+    pub num_constraints: usize,
+    /// Witness length.
+    ///
+    /// `m - l - 1` in the CCS/HyperNova papers.
+    pub num_vars: usize,
+    /// Length of the public input `X`. It is expected to have a leading
+    /// `ScalarField::ONE` element, thus this field must be non-zero.
+    ///
+    /// `l + 1`, w.r.t. the CCS/HyperNova papers.
+    pub num_io: usize,
+    /// Number of matrices.
+    ///
+    /// `t` in the CCS/HyperNova papers.
+    pub num_matrices: usize,
+    /// Number of multisets.
+    ///
+    /// `q` in the CCS/HyperNova papers.
+    pub num_multisets: usize,
+    /// Max cardinality of the multisets.
+    ///
+    /// `d` in the CCS/HyperNova papers.
+    pub max_cardinality: usize,
+    /// Set of constraint matrices.
+    pub Ms: Vec<SparseMatrix<G::ScalarField>>,
+    /// Multisets of selector indices, each paired with a constant multiplier.
+    pub cSs: Vec<(G::ScalarField, Vec<usize>)>,
+}
+
+impl<G: CurveGroup> CCSShape<G> {
+    /// Checks if the CCS instance together with the witness `W` satisfies the CCS constraints determined by `shape`.
+    pub fn is_satisfied<C: PolyCommitmentScheme<G>>(
+        &self,
+        U: &CCSInstance<G, C>,
+        W: &CCSWitness<G>,
+        ck: &C::PolyCommitmentKey,
+    ) -> Result<(), Error> {
+        assert_eq!(W.W.len(), self.num_vars);
+        assert_eq!(U.X.len(), self.num_io);
+
+        let z = [U.X.as_slice(), W.W.as_slice()].concat();
+        let Mzs: Vec<Vec<G::ScalarField>> = ark_std::cfg_iter!(&self.Ms)
+            .map(|M| M.multiply_vec(&z))
+            .collect();
+
+        let mut acc = vec![G::ScalarField::ZERO; self.num_constraints];
+        for (c, S) in &self.cSs {
+            let mut hadamard_product = vec![*c; self.num_constraints];
+
+            for idx in S {
+                ark_std::cfg_iter_mut!(hadamard_product)
+                    .enumerate()
+                    .for_each(|(j, x)| *x *= Mzs[*idx][j]);
+            }
+
+            ark_std::cfg_iter_mut!(acc)
+                .enumerate()
+                .for_each(|(i, s)| *s += hadamard_product[i]);
+        }
+
+        if ark_std::cfg_iter!(acc).any(|s| !s.is_zero()) {
+            return Err(Error::NotSatisfied);
+        }
+
+        if U.commitment_W != W.commit::<C>(ck) {
+            return Err(Error::NotSatisfied);
+        }
+
+        Ok(())
+    }
+
+    pub fn is_satisfied_linearized<C: PolyCommitmentScheme<G>>(
+        &self,
+        U: &LCCSInstance<G, C>,
+        W: &CCSWitness<G>,
+        ck: &C::PolyCommitmentKey,
+    ) -> Result<(), Error> {
+        assert_eq!(W.W.len(), self.num_vars);
+        assert_eq!(U.X.len(), self.num_io);
+
+        let z = [U.X.as_slice(), W.W.as_slice()].concat();
+        let Mzs: Vec<G::ScalarField> = ark_std::cfg_iter!(&self.Ms)
+            .map(|M| vec_to_mle(M.multiply_vec(&z).as_slice()).evaluate::<G>(U.rs.as_slice()))
```

Comment:
To ensure endianness compatibility with sumcheck, we need to ensure that the polynomial `P(v): DenseMultilinearExtension` that we pass to `ark-ml-sumcheck` later on satisfies `P(v).evaluate(r) == vec_to_mle(v).evaluate(r)`. The `evaluate()` function for `DenseMultilinearExtension` is little-endian, i.e. the first variable is encoded by the _least significant bit_ of the index. That's the opposite of how `evaluate()` works on spartan's `DensePolynomial`, so we either need `P(v) = DenseMultilinearExtension::from_coefficients_vec(v.rev_bits())` xor we need `vec_to_mle(v) = DensePolynomial::new(v.rev_bits())`. Since both of these maps use clones anyway, it shouldn't be more expensive to reverse the bits.